File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        // dfs left tree to try to find k,
        // else look for k - (num elements in left subtree)
        int res;
        dfs(root, k, res);
        return res;
    }

    void dfs(TreeNode* root, int& k, int& res) {
        if (!root) return;
        dfs(root->left, k, res);
        k--;
        if (k == 0) {
            res = root->val;
            return;
        }
        dfs(root->right, k, res);
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.go
func productExceptSelf(nums []int) []int {
	res := make([]int, len(nums))
	left := make([]int, len(nums))
	right := make([]int, len(nums))
	left[0] = 1
	right[len(nums) - 1] = 1

	for i := 1; i < len(nums); i++ {
		left[i] = left[i - 1] * nums[i - 1]
		right[len(nums) - i - 1] = right[len(nums) - i] * nums[len(nums) - i]
	}

	for i := 0; i < len(nums); i++ {
		res[i] = left[i] * right[i]
	}

	return res
}
